drafts/sorting-data-with-clojure.md

```diff
@@ -0,0 +1,140 @@
+
+Already it is clear that within one RDBMS system we can have multiple behaviours depending on the underlying storage engine used by the table.
+
+Long story short, we have no gaurantees and this can easily nip us in the behind if we're not careful.
+
+## Leaving it to the caller
+
+Knowing now that we cannot depend on the source to sort the data for us, it is almost certainly up to the caller to sort the data. This often comes in the form of a query parameter, be it an `ORDER BY` clause in a SQL statement or a query parameter to an API.
+
+More bespoke sorting tends to happen in the consuming code, not at the source, and this is where we'll look at what Clojure offers us.
+
+# Sorting data with Clojure
+
+Clojure is great at taming data of all kinds, here I just want to explore a few ways to get some order to your data using simple functions.
+
+## Example data
+
+For the examples below I’m going to be working with some invoice data. Each invoice looks something like this:
+
+```clojure
+{:invoice/number “ACMEINV00001”
+ :invoice/date “2016-11-25”
+ :invoice/total-before-tax 100
+ :invoice/tax 10
+ :invoice/items [...]}
+```
+
+I’ll leave it up to you to imagine how rich these data structures can become in a real invoicing system.
+
+## Getting started with (sort-by)
+
+The first requirement could be to sort a vector of invoices by total. These is relatively simple with `(sort-by)`:
+
+```clojure
+(sort-by :invoice/total-before-tax invoices)
+```
+
+The first argument to `(sort-by)` should be function, and since keywords are functions of maps you can just specify the key in the map to be used. One caveat, the value of the entry in the map must be comparable.
+
+That gets you a new vector, with the smallest invoices first and the valuable ones at the end. Hardly useful for business, so lets flip it around by supplying a comparator function too:
+
+```clojure
+(sort-by > :invoice/total-before-tax invoices)
+```
+
+Now we’re cooking with gas! The most valuable invoices are now at the head of the list! Need the top 10? Just `(take 10)` and you’re set.
+
+How did this happen? Clojure compared the values returned by `:invoice/total-before-tax` using the `>` function.
+
+## Sorting by composite keys
+
+The next requirement might be to sort by `:invoice/number` and `:invoice/total-before-tax`. Imagine the idea is that when two invoices have the same total that they are then sorted by their invoice number to show some kind of implied order.
+
+This is where our friend `(juxt)` comes in. `(juxt)` accepts a list of functions and returns a new function, that when called, returns the results of all the original functions in a vector.
+
+```clojure
+(def head-and-tail (juxt [first last]))
```

Comment:
`juxt` is variadic; it takes functions directly, rather than a single arg with a seq of functions:

https://clojuredocs.org/clojure.core/juxt

`(juxt first last)`